src/biogeochem/SPMMod.F90

```diff
@@ -0,0 +1,805 @@
+  
+  !diagnoal matrix only store diagnoal entries
+
+     real(r8), pointer :: DM(:,:)         ! entries in diagonal matrix (unit,diagonal matrix index)
+     integer SM                           ! Size of matrix, eg. for nxn matrix, SM=n
+     integer num_unit                     ! number of active unit, such as patch, col, or gridcell
+     integer begu                         ! begin index of unit in current process
+     integer endu                         ! end index of unit in current process
+  
+  contains
+
+    procedure, public :: InitDM           ! subroutine to initialize diagonal matrix type
+    procedure, public :: SetValueDM       ! subroutine to set values in diagonal matrix
+
+  end type diag_matrix_type
+
+  type, public :: vector_type
+  
+  !vector 
+
+     real(r8), pointer :: V(:,:)          ! entries in vector (unit,vector index)
+     integer SV                           ! Size of vector
+     integer num_unit                     ! number of active unit, such as patch, col, or gridcell
+     integer begu                         ! begin index of unit in current process
+     integer endu                         ! end index of unit in current process
+
+  contains 
+  
+    procedure, public :: InitV            ! subroutine to initialize vector type
+    procedure, public :: ReleaseV         ! subroutine to deallocate veector type
+    procedure, public :: SetValueV        ! subroutine to set values in vector
+    procedure, public :: SetValueV_scaler ! subroutine to set a constant value to a vector
+    procedure, public :: SPMM_AX          ! subroutine to calculate multiplication X(vector)=A(sparse matrix)*X(vector)
+
+  end type vector_type
+
+contains
+
+
+subroutine InitSM(this,SM_in,begu,endu,maxsm)
+
+! Initialize the sparse matrix by giving the boundary of landunit/gridcell/column/patch,
+! the size of matrix. Then allocate the matrix in a sparse matrix format
+
+class(sparse_matrix_type) :: this
+integer,intent(in) :: SM_in
+integer,intent(in) :: begu
+integer,intent(in) :: endu
+integer,optional,intent(in) :: maxsm
+
+this%SM = SM_in
+this%begu = begu
+this%endu = endu
+if(present(maxsm))then
+   allocate(this%M(begu:endu,1:maxsm))
+else
+   allocate(this%M(begu:endu,1:SM_in*SM_in))
+end if
+allocate(this%RI(1:SM_in*SM_in))
+allocate(this%CI(1:SM_in*SM_in))
+this%M(:,:) = -9999._r8
+this%RI(:) = -9999
+this%CI(:) = -9999
+this%NE    = -9999
+
+end subroutine InitSM
+
+
+subroutine SetValueSM(this,begu,endu,num_unit,filter_u,M,I,J,NE_in)
+
+! Set sparse matrix values by giving all non-zero values and the corresponding row and column indices.
+! The information of active landunit/gridcell/column/patch is used to save computational cost.
+
+class(sparse_matrix_type) :: this
+integer,intent(in) :: begu
+integer,intent(in) :: endu
+integer,intent(in) :: NE_in
+integer,intent(in) :: num_unit
+integer,intent(in) :: filter_u(:)
+real(r8),dimension(:,:),intent(in) :: M(begu:endu,1:NE_in)
```

Comment:
The nag compiler is finding array size issues with arrays in this module. A better way to handle this is to pass arrays as M(:,:) and then check the size.